designs/drift-hash-versioning.md

```diff
@@ -0,0 +1,155 @@
+# Karpenter Drift Hash Versioning with CRD
+
+## Summary of Drift Hashing Mechanism
+
+Drift utilizes a hashing mechanism to identify when a `NodeClaim` has been drifted from its owning `NodePool`. Karpenter will hash fields that are within the `NodePool.Spec.Template`, excluding [special cases of drift](https://karpenter.sh/docs/concepts/disruption/#special-cases-on-drift) or [behavior fields](https://karpenter.sh/docs/concepts/disruption/#behavioral-fields). The `NodePool` will be annotated with the hash. When a `NodeClaim` is created, the `NodePool` hash will be annotated on the `NodeClaim`. More details can be found in the [drift documentation](https://karpenter.sh/docs/concepts/disruption/#drift). 
+
+When a `NodePool` is updated, a new hash will be annotated on the `NodePool`. The `NodeClaim` disruption controller will periodically check if the `NodePool` hash match to determine when a `NodeClaim` is drifted.
+
+## Background
+
+Karpenter's static drift mechanism works well when adding new fields to the CRD that don't have defaults, but breaks down if we add a field that has a set default, remove a field from the CRD, or set a new default for an existing field. If a cloud provider makes one of these changes to their CRDs, the static drift would change, causing all nodes to roll when the CRD is upgraded. This causes problems for cloud providers when maintainers need to make these kinds of changes. We currently have three issue that need a drift hash versioning improvement: 
+
+1. The team is looking to set cloud provider defaults for the: `nodeClassRef.apiVersion` and `nodeClassRef.kind` fields [#909](https://github.com/kubernetes-sigs/karpenter/issues/909)
+ -  **We will need to coordinate with the Azure team to make sure that `Kind` and `apiVersion` are also set on the (karpenter-provider-azure)[https://github.com/Azure/karpenter-provider-azure]**
+2. The aws team is blocked from fixing issue [#5447](https://github.com/aws/karpenter-provider-aws/issues/5447)
+3. The fix to the issue is adding volumeSize in the hashing mechanism of the `EC2NodeClass` (#5454)[https://github.com/aws/karpenter-provider-aws/pull/5454]
+
+However, this will cause all `NodeClaims` to be drifted as soon as a customer upgrades their Karpenter CRDs in their cluster. These two issues raise the need for Karpenter to understand different versions of the CRDs during a drift evaluation. The problem here is that a change on the Karpenter CRDs will cause all nodes on a customer’s cluster will be drifted without any changes to the node configuration.
+
+More generally, these are breaking changes to the hashing function that should not induce drift:
+
+1. A `NodePool` adds default values to an existing field
```

Comment:
You stated these above (and I really like the way that you have categorized them and put them into a bulleted list). I wonder if there is a way to shift things around so that you don't end up stating the same set of tenets twice 🤔

Reply:
The statement from above